torch/nn/utils/clip_grad.py

```diff
@@ -31,17 +36,29 @@ def clip_grad_norm_(parameters: _tensor_or_tensors, max_norm: float, norm_type:
         return torch.tensor(0.)
     device = parameters[0].grad.device
     if norm_type == inf:
-        total_norm = max(p.grad.detach().abs().max().to(device) for p in parameters)
+        norms = [p.grad.detach().abs().max().to(device) for p in parameters]
+        if len(norms) == 1:
+            total_norm = norms[0]
+        else:
+            total_norm = torch.max(torch.stack(norms))
     else:
         total_norm = torch.norm(torch.stack([torch.norm(p.grad.detach(), norm_type).to(device) for p in parameters]), norm_type)
+    if error_if_nonfinite and (total_norm.isnan() or total_norm.isinf()):
+        raise RuntimeError(
+            f'The total norm of order {norm_type} for gradients from '
+            '`parameters` is non-finite, so it cannot be clipped. To disable '
+            'this error and zero out gradients for this case, set '
```

Comment:
Well, I just realized I led you wrong with this updated message. The gradients are only zeroed out for the `isinf()` case and not for the `isnan()` case. I guess the error message here and in the C++ implementation should indicate that correctly. Sorry about that!

Reply:
Ah right, that makes sense

Reply:
Actually, it's not zeroed out if one of the grad elements is inf, because `inf / inf == nan`. It would only be zeroed out if the total norm is inf and all elements are finite.

```python
>>> a = torch.ones(10, requires_grad=True)
>>> a[0].mul(float('inf')).sum().backward()
>>> torch.nn.utils.clip_grad_norm_(a, 1, error_if_nonfinite=False)
tensor(inf)
>>> a.grad
tensor([nan, 0., 0., 0., 0., 0., 0., 0., 0., 0.])

```

How about we just say "To disable this error and scale the gradients with the non-finite norm anyway, set `error_if_nonfinite=False`"

Reply:
I went ahead and made that change. Let me know if something else would be better

Reply:
Sure that sounds good!